Feature/cam grid2grid

MarcelCaron-NOAA opened the pull request:
## Pull Request Testing ##

- [x] Describe testing already performed for this Pull Request:</br>
All steps of the snowfall job were run as jobs in the cron for months, and output stats and plots were approved by the CAM verification team leads.
- [x] Recommend testing for the reviewer(s) to perform, including the location of input datasets, and any additional instructions:</br>

This branch contains CAM snowfall verification development.  CAM snowfall comprises five stats jobs and one plots job.  

The stats jobs use cycling.  I recommend running the stats jobs in the cron like this:
```
30 0,6,12,18 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hireswarw_snowfall_stats.out {HOMEevs}/dev/drivers/scripts/cam/stats/jevs_hireswarw_snowfall_stats.sh >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswarw_snowfall_stats.out 2>&1
 30 0,6,12,18 * * * qsub -o /lfs/h2/emc/{USER}/marcel.caron/jevs_hireswarwmem2_snowfall_stats.out {HOMEevs}/dev/drivers/scripts/cam/stats/jevs_hireswarwmem2_snowfall_stats.sh >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswarwmem2_snowfall_stats.out 2>&
 30 0,6,12,18 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hireswfv3_snowfall_stats.out {HOMEevs}/dev/drivers/scripts/cam/stats/jevs_hireswfv3_snowfall_stats.sh >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hireswfv3_snowfall_stats.out 2>&1
 30 0,6,12,18 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_hrrr_snowfall_stats.out {HOMEevs}/dev/drivers/scripts/cam/stats/jevs_hrrr_snowfall_stats.sh >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_hrrr_snowfall_stats.out 2>&1
 30 0,6,12,18 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_namnest_snowfall_stats.out {HOMEevs}/dev/drivers/scripts/cam/stats/jevs_namnest_snowfall_stats.sh >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_namnest_snowfall_stats.out 2>&1
```

Replacing {USER} with your WCOSS2 username, and {HOMEevs} with your EVS PR testing directory.  Also, check each driver script (`{HOMEevs}/dev/drivers/scripts/...`) and set the following variables to your desired settings:
HOMEevs, DATA, COMOUT

The plots job can be run interactively at any time.  If you prefer to use the cron, I have the job set up like this:
`0 23 * * * qsub -o /lfs/h2/emc/stmp/{USER}/jevs_cam_snowfall_plots.out {HOMEevs}/dev/drivers/scripts/cam/plots/jevs_cam_snowfall_plots.sh >> /lfs/h2/emc/stmp/{USER}/cron.out/test_jevs_cam_snowfall_plots.out 2>&1`

Replacing `{USER}` and `{HOMEevs}` as with the stats cron jobs.  Also, open the plots driver script (`{HOMEevs}/dev/drivers/scripts/cam/plots/jevs_cam_snowfall_plots.sh`) and set COMIN to my test directory:
`export COMIN=/lfs/h2/emc/vpppg/noscrub/marcel.caron/temp/test_stats/$COMPONENT`

Then set HOMEevs, DATA, and COMOUT to your desired settings.

- [x] Has the code been checked to ensure that no errors occur during the execution? **[Yes]**

- [x] Do these updates/additions include sufficient testing updates? **[Yes]**

- [x] Please complete this pull request review by **[07/21/2023]**.</br>

## Pull Request Checklist ##

- [ ] Review the source issue metadata (required labels, projects, and milestone).
- [ ] Complete the PR description above.
- [ ] Ensure the PR title matches the feature branch name.
- [ ] Check the following:
- [ ]  Instructions provided on how to run
- [ ]  Developer's name is replaced by ${user} where necessary throughout the code
- [ ]  Check that the ecf file has all the proper definitions of variables
- [ ]  Check that the jobs file has all the proper settings of COMIN and COMOUT and other input variables
- [ ]  Check to see that the output directory structure is followed
- [ ]  Be sure that you are not using MET utilities outside the METplus wrapper structure

- [ ] After submitting the PR, select **Development** issue with the original issue number.
- [ ] After the PR is approved, merge your changes. If permissions do not allow this, request that the reviewer do the merge.
- [ ] Close the linked issue.

Commit:
Added workflow for cam plotting

Commit:
full plotting scripts integrated into cam plots workflow

Commit:
Fix typos and syntax errors + use COMIN for input stats

Commit:
Merge branch 'feature/CAM_grid2obs_priority' into feature/CAM_plotting
feature/CAM_grid2obs_priority is the next branch to merge into the
authoritative branch; therefore feature/CAM_plotting needs to stay up to
date with changes in feature/CAM_grid2obs_priority

Commit:
Adjust settings and minor bug fixes

Commit:
Improve plotting logic for unit conversion, fix bugs

Commit:
Merge branch 'develop' into feature/CAM_plotting

Commit:
changed filename structure and unit conversion logic + removed hpbl vhrmean jobs

Commit:
Fixed minor bugs + changed version and HOMEevs directory for test runs

Commit:
fix minor bugs preventing successful runs + update module versions and HOMEevs

Commit:
Merge develop

Commit:
Add NBRCNT, INTERP_PNTS, and FSS capability

Commit:
Add and adapt global_det precip spatial map plotting scripts to CAM directories

Commit:
Add capability to plot MRMS + lon sanitization + minor edits to plot specs

Commit:
add pr and hi jobs to PCPCombine job type

Commit:
Edit workflows for hiresws to include pr and hi pcp_combine jobs + include step where pcp_combine output are copied to COMOUT for all models

Commit:
fix precip plot failure + finish spatial map code + minor bugfixes

Commit:
Fix minor bugs causing errors that prevent CAM_plotting and CAM_precip jobs from completing

Commit:
Change vx_mask strings used for setting map extent

Commit:
Fix final errors in spatial mapping code + reset VDATE to ops settings + move MET/METplus loading to prod dirs

Commit:
Merge branch 'develop' into feature/CAM_plotting

Commit:
First commit

Commit:
Fix reformat and generate configurations for snowfall verif

Commit:
Add workflow for working snowfall verification jobs

Commit:
Merge with develop

Commit:
merge develop

Commit:
merge develop and remote

Commit:
Merge branch 'develop' into feature/CAM_grid2grid

Commit:
Merge develop

Commit:
Merge branch 'develop' into feature/CAM_grid2grid

Commit:
Update to new developer settings

Commit:
Merge branch 'develop' into feature/CAM_grid2grid

Commit:
Merge branch 'develop' into feature/CAM_grid2grid

Commit:
Merge branch 'develop' into feature/CAM_grid2grid

Commit:
Change HOMEevs dir for snowfall

Commit:
updated run.ver, added cycling, finished maillist restructuring, and other updates

Commit:
merge develop

Commit:
Merge develop

Commit:
Merge develop

Commit:
Switch HOMEevs to noscrub testing directory

Commit:
correct driver script extensions

Commit:
Add a toggle option for labeling the x-axis of threshmean plots with all requested thresholds, or only those corresponding to plotted data points

Commit:
Apply unit conversion to threshold values if including all requested values + changed input data directory for testing

Commit:
Merge develop

Commit:
Changed module loading procedure and input directory defs

Commit:
Merge branch 'develop' into feature/CAM_grid2grid

Commit:
Remove unecessary 'cam' conditional

Commit:
Correct spatial mapping settings

ShelleyMelchior-NOAA commented:
@MarcelCaron-NOAA  -- I have cloned your fork and checked out your branch `feature/CAM_grid2grid`.
I have modifed the stats drivers and will begin running them.
As for the plots drivers, is there a stats archive I should point to for COMIN? Or are the stats I will generate be sufficient?

ShelleyMelchior-NOAA commented on dev/drivers/scripts/cam/plots/jevs_cam_grid2obs_plots.sh:
@MarcelCaron-NOAA , do you need me to re-test this processing in this PR?

MarcelCaron-NOAA replied:
@ShelleyMelchior-NOAA Thanks for checking; no I don't think this needs to be retested.  These variables don't affect the plotting job, given they were pointing to defunct directories in the first place in the recent successful bugfix PR.  I corrected them to be consistent with other corrections elsewhere, but in a later PR I will probably have these removed and the job retested then.  If you prefer these stay the same until then, I can revert the changes.

ShelleyMelchior-NOAA replied:
I agree that the changes will not impact a successful run. You can leave as is for this PR. I will only test snowfall.

MarcelCaron-NOAA commented:
> @MarcelCaron-NOAA -- I have cloned your fork and checked out your branch `feature/CAM_grid2grid`. I have modifed the stats drivers and will begin running them. As for the plots drivers, is there a stats archive I should point to for COMIN? Or are the stats I will generate be sufficient?

@ShelleyMelchior-NOAA you can set COMIN to the following directory, which is my testing directory:
export COMIN=/lfs/h2/emc/vpppg/noscrub/marcel.caron/temp/test_stats/$COMPONENT

ShelleyMelchior-NOAA commented:
And I just read your instructions more closely and you already indicated using your stats archive for COMIN. Apologies for not seeing that the first time.

MarcelCaron-NOAA commented:
Not a problem!

ShelleyMelchior-NOAA commented:
@MarcelCaron-NOAA , the stats processing is churning away. I figured I could run plots anytime, pointing to your stats archive for COMIN. The plots processing failed, though, b/c cyc is not defined. From the log file:
`77 + export cycle=tz`
`77 + cycle=tz`
That causes `setpdy.sh` to not work.
I see cyc used in `JEVS_CAM_PLOTS` (`export cycle=t${cyc}z`), but it is not defined. I also do not see it defined in the upstream driver (`jevs_cam_snowfall_plots.sh`). I can export it from the driver. I assume 00 is OK?

For full context, the logfile is: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_snowfall_plots.o82313488`

-------

While writing this message, the stats processing completed for the 00 cron run. Please review.
COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/stats/cam`
logfiles: `/u/shelley.melchior/jevs_*_snowfall_stats.o*`

Commit:
export cyc in plotting driver scripts

ShelleyMelchior-NOAA commented:
I re-ran the plots driver setting cyc=00 and it ran to completion. Please check out the results.
COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/plots/cam/atmos`
logfile: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_snowfall_plots.o82319011`

Can you check out the format of COMOUT? The EVS para has plots in `atmos.YYYYMMDD`, not simply `atmos`.

Commit:
Change COMOUT definition in driver and J-job scripts

MarcelCaron-NOAA commented:
@ShelleyMelchior-NOAA The stats output for 00Z looks good.  All models completed cleanly and have mini stat files in COMOUT/atmos.20230712--all except the hrrr which hasn't completed yet as I'm writing.  The final stats should show up in COMOUT after the 18Z cycle completes.
 
The plots job also completed without any issue, and the output file looks good to me.

I've pushed changes to the driver script, the first is defining cyc.  COMOUT is more complicated, since I can't define a COMOUT that includes "/atmos.YYYYmmdd" without using VDATE, which is defined afterward in the J-job.  I opted to overwrite this by defining VDATE in the driver script.  Is that okay?

ShelleyMelchior-NOAA commented on versions/run.ver:
```diff
@@ -1,4 +1,6 @@
 export evs_ver=v1.0
+export met_ver=11.0.0
+export metplus_ver=5.0.0
 
```
@MarcelCaron-NOAA -- any reason whey these old versions were introduced?

MarcelCaron-NOAA replied:
No these were not intentional additions.  Removing them--thanks.

ShelleyMelchior-NOAA commented:
You are correct -- hrrr was still running when I submitted the comment, thinking you wouldn't see it until after it finished (it was running fine). But you snuck in there. lol All cron processes have been running well for 0030, 0630, and 1230. We can wait for the 1830 cron to evaluate results before approving the PR. In the mean time, I have started the code review.

ShelleyMelchior-NOAA commented on jobs/cam/plots/JEVS_CAM_PLOTS:
Since this is a shared J-job across cam, I wanted to get extra eyes on this file's changes. I think everything is OK for this iteration, but please double check me. @LoganDawson-NOAA @MarcelCaron-NOAA @BinbinZhou-NOAA

MarcelCaron-NOAA replied:
I can confirm that the changes made to JEVS_CAM_PLOTS and JEVS_CAM_STATS are changes I've intentionally made in this PR, and these changes shouldn't interfere with others' jobs.  I'll let the others confirm as well.

Commit:
Removing old met versions from run.ver

ShelleyMelchior-NOAA commented:
The 1830 cron entries just kicked off.

ShelleyMelchior-NOAA commented:
@MarcelCaron-NOAA -- all crons for today are completed. You can check the final results in COMOUT for stats.
COMOUT=`/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/stats/cam`

Please also sync this feature branch w/ auth develop. If all looks good we can merge the PR very soon.

Commit:
Merge branch 'develop' into feature/CAM_grid2grid

MarcelCaron-NOAA commented:
All of the jobs outputted final .stat files that are about the expected file size.  Each of the cron jobs (4 cycles x 5 models = 20 jobs) completed cleanly.  Manual search found no errors. These seem good to go.

I just merged the recent updates in develop.  If there are no new issues, then I am fine with merging this PR.

ShelleyMelchior-NOAA commented:
@MarcelCaron-NOAA -- the only thing outstanding is to make sure JEVS_CAM_PLOTS is indeed the latest. Since it is shared across cam I just wanted that extra confirmation.

PerryShafran-NOAA commented:
@MarcelCaron-NOAA @ShelleyMelchior-NOAA @LoganDawson-NOAA @BinbinZhou-NOAA One of the things we need to worry about is to ensure that JEVS_CAM_PLOTS works for all elements of CAM.  Because I just noticed that whatever was last put into the parallel most recently hosed the nam_firewxnest plots.  We also need to make sure that it works regardless of whatever COMIN/COMOUT is set in the driver script, because eventually those will be removed because they can't be in that script; only the jobs script can have those elements.  Thus, we should probably all agree as to what COMOUT is and where the final plots will ultimately go.

COMIN should be /lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v1.0
COMOUT should also be /lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v1.0
COMOUTplots should be /lfs/h2/emc/vpppg/noscrub/emc.vpppg/evs/v1.0/plots/cam/atmos.$VDATE

We should not have COMIN or COMOUT be different by model.  

Also - we need to ensure that all models get the dates that they need to use for plotting.  For nam_firewxnest, we need to run setpdy.sh 31 in order to get VDATE and PDYm31 to get plots.  The method to get these dates needs to be the same for all, and they need to be in the J-job and not anywhere else in the system.  

We should do this now and we should all test it, so we can have a final version that works for all and also will pass muster with NCO.  

Perry

Commit:
Change driver script names + change J-jobs and exec scripts to use COMOUTplots

Commit:
Make sure COMOUTplots will be an available directory + change plots tarball write dir to DATA, then copy to COMOUT

Commit:
Add check for COMOUTplots + correct mkdir def for COMOUTplots (use explicit def)

Commit:
Merge branch 'develop' into feature/CAM_grid2grid

Commit:
Save both ASNOW and SNOD plots as ASNOW for webpage compatibility

ShelleyMelchior-NOAA commented on dev/drivers/scripts/cam/stats/jevs_cam_hireswarwmem2_snowfall_stats.sh:
```diff
@@ -0,0 +1,62 @@
+source /usr/share/lmod/lmod/init/sh
+module reset
+source $HOMEevs/modulefiles/$COMPONENT/${COMPONENT}_${STEP}.sh
+export MET_bin_exec="bin"
```
We can remove this line throughout all driver scripts now that the setting is contained within `machine.conf`.
`export MET_bin_exec="bin"`

MarcelCaron-NOAA replied:
@ShelleyMelchior-NOAA I removed this line in all cam/precip, /grid2obs, and /snowfall driver scripts, and added the machine.conf as a config to all METplus runs to access its settings.

ShelleyMelchior-NOAA commented:
In light of the recent updates in this PR re: plots changes, I need to re-run plots. @MarcelCaron-NOAA , is there a need to re-run stats?

Also, a PR has been merged. Please update your feature branch.

ShelleyMelchior-NOAA commented:
@MarcelCaron-NOAA , I re-ran the snowfall plots. Please review the output.
COMOUT= `/lfs/h2/emc/vpppg/noscrub/shelley.melchior/evs/v1.0/plots/cam`
logfile: `/lfs/h2/emc/ptmp/shelley.melchior/EVS_out/jevs_cam_snowfall_plots.o72908677`

MarcelCaron-NOAA commented:
Hi @ShelleyMelchior-NOAA thanks for re-runnning.  Testing only the plots scripts is sufficient I think.

I checked the output logs and the job seems to have run fine.  The tar file looks good as well!

I'll work now on merging the updates from the CAM/ptype PR into this branch as well as addressing your comments here

Commit:
merge develop

Commit:
Remove driver-level MET_bin_exec def and include machine_conf config in all METplus execute commands

ShelleyMelchior-NOAA commented:
Thanks, @MarcelCaron-NOAA ! I am satisfied with all of the changes. I want express one last time to check w/ you on the two J-jobs. Are all of those changes kosher and non-impactful for the other developers in the cam component of EVS?

MarcelCaron-NOAA commented:
> Thanks, @MarcelCaron-NOAA ! I am satisfied with all of the changes. I want express one last time to check w/ you on the two J-jobs. Are all of those changes kosher and non-impactful for the other developers in the cam component of EVS?

@ShelleyMelchior-NOAA  I don't see these changes having any impact on the other cam cases.  These J-jobs are largely equivalent to the dev version.  The notable changes either impact only the jobs I'm working on, or they change settings that Logan/Perry had already suggested changing (e.g., defining VDATE using PDY and not using the bash date command).  So, barring concerns from the other developers, I think the changes are safe to merge!

ShelleyMelchior-NOAA commented:
Then let's do this! I have been doing the code review as we have been going along. I will submit the formal code review and then merge.

ShelleyMelchior-NOAA reviewed (approved):
The PR is approved for merge.
